File: ios/GeocoderFunctions.swift
import ArcGIS
import Foundation

/// Free functions backing the JS `geocoder` namespace — forward / reverse geocoding via a
/// `LocatorTask`. Registered as `AsyncFunction`s in `ExpoArcgisGeometryModule`.
///
/// SuggestResult round-trip
/// ─────────────────────────
/// `suggest(...)` attaches a `suggestionId` (Int) to each returned item and stashes the native
/// `SuggestResult` (and its locator URL) in `SuggestRegistry`. The registry is REPLACED on each
/// new `suggest` call (ids restart from 0) so memory never grows unboundedly.
/// `geocodeSuggestion(suggestionId, locatorUrl?)` looks up the held result and calls
/// `LocatorTask.geocode(forSuggestResult:)` — the SDK's precise overload that avoids a text
/// re-search.

/// Holds the native `SuggestResult`s from the most recent `suggest` call.
private final class SuggestRegistry: @unchecked Sendable {
  static let shared = SuggestRegistry()
  private let lock = NSLock()
  private var results: [Int: SuggestResult] = [:]
  private var locatorUrl: String = ""

  /// Replaces the registry with a fresh set of results (called at the start of each `suggest`).
  func reset(url: String, results: [(Int, SuggestResult)]) {
    lock.lock()
    defer { lock.unlock() }
    self.locatorUrl = url
    self.results = Dictionary(uniqueKeysWithValues: results)
  }

  func lookup(_ id: Int) -> (result: SuggestResult, locatorUrl: String)? {
    lock.lock()
    defer { lock.unlock() }
    guard let r = results[id] else { return nil }
    return (r, locatorUrl)
  }
}

private let worldGeocoderURL = "https://geocode-api.arcgis.com/arcgis/rest/services/World/GeocodeServer"

/// Caches one `LocatorTask` per service URL so repeated geocodes don't reload the service metadata.
private final class LocatorCache: @unchecked Sendable {
  static let shared = LocatorCache()
  private let lock = NSLock()
  private var tasks: [String: LocatorTask] = [:]

  func task(for locator: String) -> LocatorTask? {
    lock.lock()
    defer { lock.unlock() }
    if let task = tasks[locator] { return task }
    // A scheme'd URL (online geocode service); otherwise a local file path (an offline `.loc`).
    let parsed = URL(string: locator).flatMap { $0.scheme != nil ? $0 : nil } ?? URL(fileURLWithPath: locator)
    let task = LocatorTask(url: parsed)
    tasks[locator] = task
    return task
  }
}

func geocode(_ searchText: String, _ params: [String: Any]) async throws -> [[String: Any]] {
  guard let locator = LocatorCache.shared.task(for: locatorURL(params)) else { return [] }
  let geocodeParameters = buildGeocodeParameters(params)
  let results: [GeocodeResult]
  if let raw = params["searchValues"] as? [String: Any],
     !raw.isEmpty,
     let searchValues = raw as? [String: String] {
    results = try await locator.geocode(forSearchValues: searchValues, using: geocodeParameters)
  } else {
    results = try await locator.geocode(forSearchText: searchText, using: geocodeParameters)
  }
  return results.map(serializeGeocodeResult)
}

func reverseGeocode(_ point: [String: Any], _ params: [String: Any]) async throws -> [[String: Any]] {
  guard let locator = LocatorCache.shared.task(for: locatorURL(params)),
    let location = geometryFromDict(point) as? Point
  else { return [] }
  let results = try await locator.reverseGeocode(forLocation: location, parameters: buildReverseGeocodeParameters(params))
  return results.map(serializeGeocodeResult)
}

private func locatorURL(_ params: [String: Any]) -> String {
  params["locatorUrl"] as? String ?? worldGeocoderURL
}

private func buildGeocodeParameters(_ params: [String: Any]) -> GeocodeParameters {
  let parameters = GeocodeParameters()
  if let maxResults = params["maxResults"] as? NSNumber { parameters.maxResults = maxResults.intValue }
  if let countryCode = params["countryCode"] as? String { parameters.countryCode = countryCode }
  if let categories = params["categories"] as? [String] { parameters.addCategories(categories) }
  if let names = params["resultAttributeNames"] as? [String] { parameters.addResultAttributeNames(names) }
  if let wkid = params["outputSpatialReference"] as? NSNumber,
     let sr = SpatialReference(wkid: WKID(wkid.intValue)!) {
    parameters.outputSpatialReference = sr
  }
  if let location = (params["preferredSearchLocation"] as? [String: Any]).flatMap(geometryFromDict) as? Point {
    parameters.preferredSearchLocation = location
  }
  return parameters
}

private func buildReverseGeocodeParameters(_ params: [String: Any]) -> ReverseGeocodeParameters {
  let parameters = ReverseGeocodeParameters()
  if let maxResults = params["maxResults"] as? NSNumber { parameters.maxResults = maxResults.intValue }
  if let maxDistance = params["maxDistance"] as? NSNumber { parameters.maxDistance = maxDistance.doubleValue }
  return parameters
}

func suggest(_ searchText: String, _ params: [String: Any]) async throws -> [[String: Any]] {
  let url = locatorURL(params)
  guard let locator = LocatorCache.shared.task(for: url) else { return [] }
  let results = try await locator.suggest(forSearchText: searchText, parameters: buildSuggestParameters(params))
  // Assign a stable id to each result and stash in the registry (replaces any prior batch).
  let indexed = results.enumerated().map { (id: $0.offset, result: $0.element) }
  SuggestRegistry.shared.reset(url: url, results: indexed.map { ($0.id, $0.result) })
  return indexed.map { entry in
    var dict = serializeSuggestResult(entry.result)
    dict["suggestionId"] = entry.id
    return dict
  }
}

func geocodeSuggestion(_ suggestionId: Int, _ params: [String: Any]) async throws -> [[String: Any]] {
  // Allow the caller to override the locator URL, but fall back to the one stored by suggest().
  let overrideUrl = params["locatorUrl"] as? String
  guard let (suggestResult, storedUrl) = SuggestRegistry.shared.lookup(suggestionId) else {
    throw NSError(
      domain: "ExpoArcgis",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: "suggestionId \(suggestionId) not found — call suggest() first"]
    )
  }
  let url = overrideUrl ?? storedUrl
  guard let locator = LocatorCache.shared.task(for: url) else { return [] }
  let results = try await locator.geocode(forSuggestResult: suggestResult)
  return results.map(serializeGeocodeResult)
}

private func buildSuggestParameters(_ params: [String: Any]) -> SuggestParameters {
  let parameters = SuggestParameters()
  if let maxResults = params["maxResults"] as? NSNumber { parameters.maxResults = maxResults.intValue }
  if let categories = params["categories"] as? [String] { parameters.addCategories(categories) }
  if let location = (params["preferredSearchLocation"] as? [String: Any]).flatMap(geometryFromDict) as? Point {
    parameters.preferredSearchLocation = location
  }
  return parameters
}

func serializeSuggestResult(_ result: SuggestResult) -> [String: Any] {
  ["label": result.label, "isCollection": result.isCollection]
}

func serializeGeocodeResult(_ result: GeocodeResult) -> [String: Any] {
  [
    "label": result.label,
    "score": result.score,
    "attributes": serializeAttributes(result.attributes),
    "location": result.displayLocation.map(dictFromGeometry) ?? NSNull(),
  ]
}

File: android/src/main/java/expo/modules/arcgis/GeocoderFunctions.kt
package expo.modules.arcgis

import com.arcgismaps.geometry.Point
import com.arcgismaps.geometry.SpatialReference
import com.arcgismaps.tasks.geocode.GeocodeParameters
import com.arcgismaps.tasks.geocode.GeocodeResult
import com.arcgismaps.tasks.geocode.LocatorTask
import com.arcgismaps.tasks.geocode.ReverseGeocodeParameters
import com.arcgismaps.tasks.geocode.SuggestParameters
import com.arcgismaps.tasks.geocode.SuggestResult
import java.util.concurrent.ConcurrentHashMap

/**
 * Free functions backing the JS `geocoder` namespace — forward / reverse geocoding via a
 * [LocatorTask]. Registered as `AsyncFunction`s in `ExpoArcgisGeometryModule`.
 *
 * SuggestResult round-trip
 * ─────────────────────────
 * [suggest] attaches a `suggestionId` (Int) to each returned item and stashes the native
 * [SuggestResult] (and its locator URL) in [SuggestRegistry]. The registry is REPLACED on each
 * new [suggest] call (ids restart from 0) so memory never grows unboundedly.
 * [geocodeSuggestion] looks up the held result and calls `LocatorTask.geocode(suggestResult)` —
 * the SDK's precise overload that avoids a text re-search.
 */

/** Holds the native [SuggestResult]s from the most recent [suggest] call. */
private object SuggestRegistry {
  @Volatile private var locatorUrl: String = ""
  @Volatile private var results: Map<Int, SuggestResult> = emptyMap()

  /** Replaces the registry with a fresh set of results (called at the start of each [suggest]). */
  @Synchronized fun reset(url: String, items: List<SuggestResult>) {
    locatorUrl = url
    results = items.mapIndexed { id, r -> id to r }.toMap()
  }

  @Synchronized fun lookup(id: Int): Pair<SuggestResult, String>? =
    results[id]?.let { it to locatorUrl }
}

private const val WORLD_GEOCODER =
  "https://geocode-api.arcgis.com/arcgis/rest/services/World/GeocodeServer"

/** Caches one [LocatorTask] per service URL so repeated geocodes don't reload service metadata. */
private val locators = ConcurrentHashMap<String, LocatorTask>()

private fun locatorTask(params: Map<String, Any?>): LocatorTask {
  val url = params["locatorUrl"] as? String ?: WORLD_GEOCODER
  return locators.getOrPut(url) { LocatorTask(url) }
}

internal suspend fun geocode(searchText: String, params: Map<String, Any?>): List<Map<String, Any?>> {
  val locator = locatorTask(params)
  val parameters = buildGeocodeParameters(params)
  @Suppress("UNCHECKED_CAST")
  val searchValues = (params["searchValues"] as? Map<*, *>)
    ?.mapNotNull { (k, v) -> if (k is String && v is String) k to v else null }
    ?.toMap()
  return if (!searchValues.isNullOrEmpty()) {
    locator.geocode(searchValues, parameters).getOrThrow()
  } else {
    locator.geocode(searchText, parameters).getOrThrow()
  }.map { serializeGeocodeResult(it) }
}

internal suspend fun reverseGeocode(point: Map<String, Any?>, params: Map<String, Any?>): List<Map<String, Any?>> {
  val location = geometryFromDict(point) as? Point ?: return emptyList()
  return locatorTask(params).reverseGeocode(location, buildReverseGeocodeParameters(params)).getOrThrow()
    .map { serializeGeocodeResult(it) }
}

private fun buildGeocodeParameters(params: Map<String, Any?>): GeocodeParameters = GeocodeParameters().apply {
  (params["maxResults"] as? Number)?.toInt()?.let { maxResults = it }
  (params["countryCode"] as? String)?.let { countryCode = it }
  (params["categories"] as? List<*>)?.filterIsInstance<String>()?.let { categories.addAll(it) }
  (params["resultAttributeNames"] as? List<*>)?.filterIsInstance<String>()?.let { resultAttributeNames.addAll(it) }
  (params["outputSpatialReference"] as? Number)?.toInt()?.let { outputSpatialReference = SpatialReference(it) }
  ((params["preferredSearchLocation"] as? Map<*, *>)?.let { geometryFromDict(it) } as? Point)
    ?.let { preferredSearchLocation = it }
}

private fun buildReverseGeocodeParameters(params: Map<String, Any?>): ReverseGeocodeParameters =
  ReverseGeocodeParameters().apply {
    (params["maxResults"] as? Number)?.toInt()?.let { maxResults = it }
    (params["maxDistance"] as? Number)?.toDouble()?.let { maxDistance = it }
  }

internal suspend fun suggest(searchText: String, params: Map<String, Any?>): List<Map<String, Any?>> {
  val url = params["locatorUrl"] as? String ?: WORLD_GEOCODER
  val results = locatorTask(params).suggest(searchText, buildSuggestParameters(params)).getOrThrow()
  // Stash all results in the registry (replaces any prior batch; ids restart from 0).
  SuggestRegistry.reset(url, results)
  return results.mapIndexed { id, r ->
    serializeSuggestResult(r) + mapOf("suggestionId" to id)
  }
}

internal suspend fun geocodeSuggestion(suggestionId: Int, params: Map<String, Any?>): List<Map<String, Any?>> {
  val (suggestResult, storedUrl) = SuggestRegistry.lookup(suggestionId)
    ?: error("suggestionId $suggestionId not found — call suggest() first")
  // Allow the caller to override the locator URL, but fall back to the one stored by suggest().
  val url = params["locatorUrl"] as? String ?: storedUrl
  return locators.getOrPut(url) { LocatorTask(url) }
    .geocode(suggestResult).getOrThrow()
    .map { serializeGeocodeResult(it) }
}

private fun buildSuggestParameters(params: Map<String, Any?>): SuggestParameters = SuggestParameters().apply {
  (params["maxResults"] as? Number)?.toInt()?.let { maxResults = it }
  (params["categories"] as? List<*>)?.filterIsInstance<String>()?.let { categories.addAll(it) }
  ((params["preferredSearchLocation"] as? Map<*, *>)?.let { geometryFromDict(it) } as? Point)
    ?.let { preferredSearchLocation = it }
}

internal fun serializeSuggestResult(result: SuggestResult): Map<String, Any?> =
  mapOf("label" to result.label, "isCollection" to result.isCollection)

internal fun serializeGeocodeResult(result: GeocodeResult): Map<String, Any?> = mapOf(
  "label" to result.label,
  "location" to result.displayLocation?.let { dictFromGeometry(it) },
  "score" to result.score,
  "attributes" to serializeAttributes(result.attributes),
)
